src/plugins/matrixops/hsplit_operation.cpp

```diff
@@ -0,0 +1,385 @@
+                {
+                    for (int i = 0; i < v.size(); i++)
+                    {
+                        if (static_cast<std::size_t>(v[i]) > num_cols)
+                            v[i] = num_cols;
+                    }
+
+                    // Include from 0 to first index
+                    ranges.push_back(std::make_pair<std::size_t, std::size_t>(
+                        std::size_t(0), std::size_t(v[0])));
+
+                    // Make a list of pairs from the passed index list
+                    for (int i = 0; i < v.size() - 1; i++)
+                    {
+                        ranges.push_back(std::make_pair<std::size_t,
+       std::size_t>(
+                            std::size_t(v[i]), std::size_t(v[i + 1])));
+                    }
+
+                    // Include from last index to last row
+                    ranges.push_back(std::make_pair<std::size_t, std::size_t>(
+                        std::size_t(v[v.size() - 1]), std::size_t(num_rows)));
+                }
+                else
+                {
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "hsplit_operation::eval",
+                        generate_error_message(
+                            "the hsplit_operation primitive requires all "
+                            "index vectors to be of size at least 1"));
+                }
+            }
+        }
+
+        primitive_argument_type hsplit_operation::hsplit2d(
+            primitive_arguments_type&& args) const
+        {
+            switch (extract_common_type(args[0]))
+            {
+            case node_data_type_bool:
+                return hsplit1d_helper<std::uint8_t>(std::move(args));
+
+            case node_data_type_int64:
+                return hsplit1d_helper<std::int64_t>(std::move(args));
+
+            case node_data_type_double:
+                return hsplit1d_helper<double>(std::move(args));
+
+            default:
+                break;
+            }
+
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "phylanx::execution_tree::primitives::"
+                    "hsplit_operation::hsplit1d",
+                generate_error_message(
+                    "the hsplit_operation primitive requires for all arguments
+       to "
+                    "be numeric data types"));
+        }
+    */
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T>
+    primitive_argument_type
+    hsplit_operation::hsplit2d_helper(primitive_arguments_type &&args) const {
+      std::array<std::size_t, PHYLANX_MAX_DIMENSIONS> matrix_dims =
+          extract_numeric_value_dimensions(args[0], name_, codename_);
+
+      std::size_t num_rows = matrix_dims[0];
+      std::size_t num_cols = matrix_dims[1];
+
+      std::size_t block_dims =
+          extract_numeric_value_dimension(args[1], name_, codename_);
+
+      std::vector<std::pair<std::size_t, std::size_t> > ranges;
```

Comment:
Is therer a way to know how many elements will be added to this vector? If yes, please call `ranges.reserve(…)` before adding elements.